src/loki/reach.cc

```diff
@@ -110,17 +131,117 @@ directed_reach SimpleReach(const DirectedEdge* edge,
         continue;
       const auto* node = tile->node(edge.endnode());
       const auto* opp_edge = tile->directededge(node->edge_index() + edge.opp_index());
+
+      // NOTE: we can go through the end of the restriction because only the start would mark a
+      // potential stopping point (maybe a path followed the restriction). We also have to stop
+      // at the start of a simple restriction because it could have been on our path
+
       // if this opposing edge is traversable we enqueue its begin node
-      if (edge_filter(opp_edge) > 0)
+      if (edge_filter(opp_edge) > 0 && !opp_edge->start_restriction() && !opp_edge->restrictions())
         enqueue(edge.endnode());
     }
   }
   // settled nodes + will be settled nodes - duplicated transitions nodes
   reach.inbound =
-      std::min(static_cast<uint32_t>(queue.size() + done.size() - transitions), max_reach);
+      std::min(static_cast<uint32_t>(queue_.size() + done_.size() - transitions_), max_reach);
+
+  // if we didnt make the limit we should do an exact check for whichever directions we didnt meet
+  auto retry_direction = (direction & kOutbound && reach.outbound < max_reach ? kOutbound : 0) |
```

Comment:
then we fall back to exact reach if our reach criteria werent met in either of the directions (in/out). and actually we only do exact reach on those directions which didnt meet the criteria to do as little exact expansion as possible.